### PR TITLE
Add Aider-inspired RepoMap

### PR DESCRIPTION
**Pull request by ryanhoangt**
This PR is to implement Aider-inspired `RepoMap` for OpenDevin's agents. The current way to construct repo map is to consider: all the files tracked by git, the relative files mentioned in the message history, and entities such as class name.

## Usage
- Set `ENABLE_REPOMAP = True` in `agenthub/codeact_agent/codeact_agent.py`. Default to `False`.
- To exclude some files, create a `.aiderignore` file in the root of the workspace. This can help accelerate the construction of the repo map (hence reducing waiting time and context cost). For instance the content of `.aiderignore` when working with OpenDevin codebase could be:
```
evaluation/
frontend/
docs/
tests/
```

## Done
- [x] add aider's [repomap.py](https://github.com/paul-gauthier/aider/blob/main/aider/repomap.py) into OpenDevin. 
- [x] make it possible to import the repomap into one of our competitive agents, such as CodeAct. This functionality should be optional, like the github message ([details](https://github.com/OpenDevin/OpenDevin/blob/f9c7c3a5205d294506a7b190b7621c0216239a3b/agenthub/codeact_agent/codeact_agent.py#L111-L114)).
- [x] refactor and remove redundant code?
- [x] Add support for `.aiderignore` to exclude some files from the git repo when building the repo map (in case the repo contains too many files).
- [x] run experiments on SWE-Bench to see if this functionality improves accuracy. Seems like this simple integration does not help much for the evaluation. We can improve it in future PRs.

Screenshot 2024-06-07 at 15 15 12

### CODE REVIEW

**Comment by ryanhoangt** [comment]
Using `RepoMap` seems to save us a few `agentskills` actions, as the agent could understand the files/classes directly from the repo map summary.

Screenshot 2024-06-04 at 20 20 19

**Comment by ryanhoangt** [comment]
Hi @neubig, can you have a look and give some reviews if this impl is aligned with what you're thinking of?

Some questions:
- Currently, I assume the workspace is a git repo and the repo map is built only from the tracked files (this is also what the authors do for `aider`). But I'm not sure what to do with the workspace that isn't a git repo. If we consider all files there will be cache, node_modules, etc
- When `aider` builds a repo map, it excludes the files that are added to the chat (since the files' content is already there). But in OpenDevin we don't have files added to chat, so maybe we can simply remove this?

**Comment by neubig** [comment]
Hey @ryanhoangt , this is awesome!

> Currently, I assume the workspace is a git repo and the repo map is built only from the tracked files (this is also what the authors do for aider). But I'm not sure what to do with the workspace that isn't a git repo. If we consider all files there will be cache dir, node_modules, etc

Maybe we can, for now, limit the use of RepoMap to github repos. Most big codebases where the repomap will be really helpful will likely be in github repos.

> When the workspace is large e.g. with OpenDevin code base, the construction of repomap seems to delay the ux quite a bit and also increase the cost. Not sure if we can improve this somehow?

Maybe we can make construction of the repomap configurable for now, and open up an issue noting that we'd like to optimize it?

> When aider builds a repo map, it excludes the files that are added to the chat (since the files' content is already there). But in OpenDevin we don't have files added to chat, so maybe we can simply consider all files?

Yeah, maybe consider all files

File: opendevin/indexing/repomap/repomap.py
Comment (neubig) [inline opendevin/indexing/repomap/repomap.py:3]: Make sure you maintain the Apache license at the top of this file attributing the original creator.

**Comment by tobitege** [comment]
How about putting Aider's "library" files into opendevin\runtime\aider and only the RepoMap specific files into its own folder?

I'm currently working on a PR for using Aider's Linter class for agentskills and my structure thus far would look like this:
grafik

**Comment by neubig** [comment]
Hey @tobitege , thanks for the discussion! I agree that maintaining a coherent directory structure will be very valuable.

Personally I'd prefer to not isolate the `aider` files by their library of origin (`aider`) but rather by their functionality (e.g. for `linter` probably something like `program_analysis`), because that would allow us to further branch out to other varieties of tools in the same general genre as necessary.

File: tests/unit/test_repomap.py
Comment (ryanhoangt) [reply tests/unit/test_repomap.py:11]: > Do we need to put in API keys in order to run these test?

I think no, we don't use any completion for these tests, just tokenization.

**Comment by neubig** [comment]
Hi @ryanhoangt , thanks a bunch for this! I can try to run a SWE-Bench Lite evaluation to see the effect of this change!

**Comment by ryanhoangt** [comment]
> Hi @ryanhoangt , thanks a bunch for this! I can try to run a SWE-Bench Lite evaluation to see the effect of this change!
> 
> I will do that after #2085 is finished.

That would be great, thank you!

File: opendevin/indexing/repomap/LICENSE.txt
Comment (li-boxuan) [inline opendevin/indexing/repomap/LICENSE.txt:99]: Did we change any file?

Comment (ryanhoangt) [reply opendevin/indexing/repomap/LICENSE.txt:99]: I'm not sure about this `LICENSE` file, I just copy from the original `aider` repo. But this `repomap.py` file is modified to remove redundant stuff like `io` object, formatting, files added to chat, and adapt some classes to be OpenDevin-specific.

Comment (li-boxuan) [reply opendevin/indexing/repomap/LICENSE.txt:99]: Yeah my point is, according to this license, we need to state that the files have been changed. Maybe as a header to each file?

Comment (ryanhoangt) [reply opendevin/indexing/repomap/LICENSE.txt:99]: Added, thank you!

**Comment by neubig** [comment]
@ryanhoangt : it seems we have a conflict in pyproject.toml, could you resolve this?

File: tests/unit/test_repomap.py
Comment (yufansong) [inline tests/unit/test_repomap.py:43]: Do we also need to `del self.repo_map` in codeact agent?

File: agenthub/codeact_agent/codeact_agent.py
Comment (yufansong) [inline agenthub/codeact_agent/codeact_agent.py:256]: Could you please share me a example of the output of `self.repo_map.get_history_aware_repo_map` and `repo_map.get_repo_map`?

And I am always confused about how `latest_user_message` get used in agent? It seems not used in anywhere. Hope someone can explain it to me. Thanks!

**Review by yufansong — COMMENTED** [review]
General LGTM, left some questions.

File: agenthub/codeact_agent/codeact_agent.py
Comment (ryanhoangt) [reply agenthub/codeact_agent/codeact_agent.py:256]: > Could you please share me a example of the output of `self.repo_map.get_history_aware_repo_map` and `repo_map.get_repo_map`?

`RepoMap.get_history_aware_repo_map` is like a wrapper that does some preprocessing (e.g. takes in the messages in the history and extracts identifiers, file paths...) and then call to `RepoMap.get_repo_map`. So their outputs should be the same and looks something like below. Not sure if this answers your question or you're talking about something else?

Screenshot 2024-06-17 at 12 07 08

> And I am always confused about how `latest_user_message` get used in agent? It seems not used in anywhere. Hope someone can explain it to me. Thanks!

It is used to modify the last message in the `messages` variable that we're going to send to the LLM.

File: tests/unit/test_repomap.py
Comment (ryanhoangt) [reply tests/unit/test_repomap.py:43]: Not sure if it is needed, as I don't see the deletion in the main script of `aider`, only in unit tests

**Comment by ryanhoangt** [comment]
> @ryanhoangt : it seems we have a conflict in pyproject.toml, could you resolve this? Once you do I'll try to run an eval.

Done it. Also I'm not quite familiar with the `swe-bench` (e.g. workspace mount things), so there will be some modifications needed for `repomap` to be able to read the files in the workspace (I tested and sometimes the repomap fails to read the files, did some ad-hoc fixes). It would be great if someone could help with this 😁

File: agenthub/codeact_agent/codeact_agent.py
Comment (yufansong) [reply agenthub/codeact_agent/codeact_agent.py:247]: > Not sure if this answers your question or you're talking about something else?

Thanks, that helps.

File: tests/unit/test_repomap.py
Comment (yufansong) [reply tests/unit/test_repomap.py:43]: So if you don't `del repo_map`, what `Windows` error will raise? Could you please explain more?

**Comment by xingyaoww** [comment]
@ryanhoangt Thanks a ton for your contribution here!! I'll take a look at this the next 24 hours. I will probably push some edits to this branch if those are needed to run SWE-Bench :)

**Comment by xingyaoww** [comment]
@ryanhoangt  Ughh, maybe it is not a good idea for me to directly push to your branch (maybe make things more messy) - feel free to revert if you want!

So I:
- make `ENABLE_REPOMAP` a configurable thing in the config
- and add `repomap_workspace`, which will default to `workspace_base` as you described earlier
- did some tweak in the SWE-Bench `run_infer.py` to set `repomap_workspace` to the corresponding github repo directory (e.g., `/workspace/XXX`) instead of `/workspace`.

You can run it by set your `config.toml`:

```
[core]
cache_dir = "/tmp/swe_bench_eval_cache_root"
workspace_base = "/home/xingyaow/OpenDevin-dev/workspace"
workspace_mount_path = "/home/xingyaow/OpenDevin-dev/workspace"
sandbox_container_image = "ghcr.io/opendevin/sandbox:main"
sandbox_type = "ssh"
ssh_hostname = "localhost"
sandbox_timeout = 120
max_chars = 100_000_000

# SWEBench eval specific
use_host_network = false
run_as_devin = false
enable_auto_lint = true
persist_sandbox = false
enable_repomap = true

[eval_gpt4o_2024_05_13]
temperature = 0.0
model = "gpt-4o-2024-05-13"
api_key = "sk-proj-YOURKEY"

```

However, there are some issues with file permissions.
- By default, SWE-Bench does not mount the host workspace into the docker container since it already has everything (e.g., testbed, environment) built into it. However, the current implementation of `RepoMap` requires the **backend** to access the workspace directory. This makes supporting the current SWE-Bench eval very challenging.
- To overcome this, I added a workaround (https://github.com/OpenDevin/OpenDevin/pull/2248/commits/7e718d77855841056d0ffc741336c785b87f0dba) to support mounting the host directory into containers - But since inside the container, the agent is working on SWE-Bench problem as "root," so it will create the `/workspace/reponame_repoversion` folder with the root user.
- And on the host machine, the `RepoMap` code will try to create files under `/workspace/reponame_repoversion` - which will cause permission issue: You cannot create something under a directory owned by root.

I'm still thinking about a workaround, but I guess long-term it will be nice if we can put this RepoMap utility into `od-runtime-cli` proposed in https://github.com/OpenDevin/OpenDevin/issues/2404 so that the RepoMap can be constructed within the runtime and passed back to the backend -- This will save us from wrestling with file permissions. cc @yufansong .

**Comment by xingyaoww** [comment]
I made more progress here by hacking with file ownership/group: https://github.com/OpenDevin/OpenDevin/pull/2490

But I think this is not the cleanest way to do this, AND when the agent edited the file / create a new file, the permission messed up again and RepoMap failed to construct again..

@ryanhoangt Do you think if is easier if we pack our current `RepoMap` into an `agentskills` to be run inside sandbox, so that we can avoid most of these issues?

**Comment by ryanhoangt** [comment]
> @ryanhoangt Do you think if is easier if we pack our current `RepoMap` into an `agentskills` to be run inside sandbox, so that we can avoid most of these issues?

Thank you for your help! I'm also planning to do this, can give it a try when I have some capacity

**Comment by xingyaoww** [comment]
>  we'll need to adjust the prompt to make the agent call the function when needed right?

Yes - or we can simulate the execution by calling that `agentskills` *for* the agent in a certain way (e.g., issuing an `IPythonRunCellAction` for every turn and get a repo map to put into user message) - it will be much clearer how we can do this once we finish the arch refractor https://github.com/OpenDevin/OpenDevin/issues/2404

> So with your current integration, can we run some instances to quickly evaluate the effectiveness of injecting RepoMap in every step? (which is different when we move it to agentskills I think)

Not really.. So the issue is that the agent is running as `root` inside the container, and the current RepoMap implementation operates on the mounted workspace (`/workspace`) where the agent creates/edits files as `root.`
In https://github.com/OpenDevin/OpenDevin/pull/2490, I added some hacks to make all the files in `/workspace` be owned by a group that contains both `root` and the host user -- so that the host user can also read and create files there.
However, Once an agent touches a file, it may change the ownership of that file back to `root` again, causing the backend on the host to have permission errors in reading that file. So, the repo map may not be correctly generated AFTER the first step.

**Comment by ryanhoangt** [comment]
> Yes - or we can simulate the execution by calling that agentskills for the agent in a certain way (e.g., issuing an IPythonRunCellAction for every turn and get a repo map to put into user message) - it will be much clearer how we can do this once we finish the arch refractor https://github.com/OpenDevin/OpenDevin/issues/2404

Okay this sounds good to me!

**Comment by neubig** [comment]
Hi! I set up my environment to run SWE-bench and would be able to evaluate this, but it seems that we should probably resolve the issue of sandboxing first.

I strongly agree that the repomap should be run inside the sandbox as the user. If it's run outside of the sandbox it could also potentially be a security risk (e.g. be used to map the host filesystem).

**Review by neubig — COMMENTED** [review]
Hi! I set up my environment to run SWE-bench and would be able to evaluate this, but it seems that we should probably resolve the issue of sandboxing first.

I strongly agree that the repomap should be run inside the sandbox as the user. If it's run outside of the sandbox it could also potentially be a security risk (e.g. be used to map the host filesystem).

Please re-request my review when it's ready!

**Comment by paul-gauthier** [comment]
Glad to see you've been inspired by aider's repo map. Though this doesn't seem to be an "aider inspired repomap", so much as a copy of aider's repomap implementation.

I see that you've taken some steps to try and comply with aider's Apache2 licence, which is appreciated. I think you are missing attribution of the source of the Apache2 code you're using. But perhaps I missed it? Probably you just need to include attribution to aider to in the comment block at the top of each file that contains derivative code.

It might be nice to also attribute the origin of all the `.scm` files in the `queries` dir. I originally made this attribution in my [repomap blog post](https://aider.chat/2023/10/22/repomap.html#credits). I just added a [README](https://github.com/paul-gauthier/aider/blob/main/aider/queries/README.md), which you might consider adapting.

**Comment by neubig** [comment]
Hi @paul-gauthier , thanks for pointing this out and thanks for your great work on Aider! We'll double-check that all of the derivative files have licenses, and can ping you before we merge this in if you'd be interested.

**Comment by ryanhoangt** [comment]
I have moved `RepoMap` under `agentskills` and tested it with a few `swe-bench` instances. It seems like there isn't much difference. The `RepoMap` constructed currently is static, and imo, it's sufficient to just execute one `get_repomap` skill at the beginning of the task for now. After #2404, we can make it dynamic by considering message history. I'm thinking about making it possible to run the skill with subfolders inside a git repo and possibly implementing some improvements in future PRs.

File: tests/unit/test_repomap.py
Comment (tobitege) [reply tests/unit/test_repomap.py:44]: We don't run in Windows anyways, right?

File: tests/unit/test_repomap.py
Comment (tobitege) [inline tests/unit/test_repomap.py:1]: Is this a unit test by or based on work by Aider? If so, what is needed here per Apache license?

File: opendevin/core/config.py
Comment (tobitege) [inline opendevin/core/config.py:197]: Please add to docstring as well.

**Review by tobitege — CHANGES_REQUESTED** [review]
Nice job! 
But 2 items I'd like to ask for your feedback (docstring, unit test license).

File: tests/unit/test_repomap.py
Comment (ryanhoangt) [reply tests/unit/test_repomap.py:1]: Yes, it's copied from Aider. We'll need to look into the license for the whole PR after other things are okay

**Comment by ryanhoangt** [comment]
> We don't run in Windows anyways, right?
@yufansong also asked this in the comments above, but I'm not sure if it's needed here. I'll remove it for now, and we can add it back if needed

**Comment by tobitege** [comment]
Not sure what the plan is, but this PR would also need integration tests like all other parts of agentskills.
Which can be quite some work due to how those are set up, run and potentially regenerated.

**Comment by ryanhoangt** [comment]
> Not sure what the plan is, but this PR would also need integration tests like all other parts of agentskills.
Which can be quite some work due to how those are set up, run and potentially regenerated.

Is it a strict requirement? If not, I think we can make this skill an experiment for now and add tests when we have a stable version that works effectively?

**Comment by tobitege** [comment]
> Is it a strict requirement? If not, I think we can make this skill an experiment for now and add tests when we have a stable version that works effectively?

Since its disabled by default, I guess it could be done later, but I definitely defer to the team for that.
To me, the license attributions are the main open issue.

**Comment by tobitege** [comment]
@ryanhoangt would you mind running `poetry lock --no-update` in your branch and commit the updated `poetry.lock` file to see if some of the failed CI actions get fixed by it?

**Comment by xingyaoww** [comment]
Agree with @tobitege here: I think we need to add attributions - Once that's done, i can try to run a full SWE bench lite to test the quality and hopefully we can get this merged. We can worry about the "dynamic" repo map later.

Or if we cannot get filepaths ready, can we make `repomap` something like a `analyze_file` function, where the agent can provide any file/folder path, and the repomap will parse and get all the dependencies? wdyt? @ryanhoangt

**Comment by ryanhoangt** [comment]
> Or if we cannot get filepaths ready now for dynamic repo map, can we make repomap something like a analyze_file function, where the agent can provide any file/folder path, and the repomap will parse and get all the dependencies?

Yes - that's what I'm planning to do in upcoming PRs - for now it can only operate at the root of a git repo

**Comment by ryanhoangt** [comment]
> would you mind running poetry lock --no-update in your branch and commit the updated poetry.lock file to see if some of the failed CI actions get fixed by it?

Could you help me run the CI again?

**Comment by tobitege** [comment]
`ERROR:root:<class 'ModuleNotFoundError'>: No module named 'git'` 🤔

**Comment by ryanhoangt** [comment]
> ERROR:root:<class 'ModuleNotFoundError'>: No module named 'git'

This one's weird, when setting up the plugin it should install `gitpython` already

**Comment by tobitege** [comment]
> > ERROR:root:<class 'ModuleNotFoundError'>: No module named 'git'
> 
> This one's weird, when setting up the plugin it should install `gitpython` already

Would it make a difference, to split that ultra-long line in the plugin setup into 2 separate commands?

**Comment by tobitege** [comment]
ohh, you have to put the packages into poetry.lock after all, if I understand this error right:
https://github.com/OpenDevin/OpenDevin/actions/runs/9640382845/job/26584274742?pr=2248
the test is importing all the agents, thus also the agentskills via CodeActAgent, but that doesn't run the setup first

**Comment by tobitege** [comment]
Sweet, the "Run e2e test with dummy agent / test (pull_request)" finally made it!

**Comment by tobitege** [comment]
```
File "/app/opendevin/runtime/plugins/agent_skills/repomap/repomap.py", line 15, in <module>
#34 4.574     import git
  File "/app/.venv/lib/python3.12/site-packages/git/__init__.py", line 298, in <module>
#34 4.574     raise ImportError("Failed to initialize: {0}".format(_exc)) from _exc
#34 4.574 
#34 4.574 ERROR:root:<class 'ImportError'>: Failed to initialize: Bad git executable.
```

**Comment by ryanhoangt** [comment]
> #34 4.574 ERROR:root:<class 'ImportError'>: Failed to initialize: Bad git executable.

Can you run again? Also I'm not sure if the unit test failure for macos (seg fault) is intermittent

**Comment by tobitege** [comment]
Hello @li-boxuan would you mind taking a look at the failed test?
I can't make sense of it right now.

**Comment by tobitege** [comment]
Hello @ryanhoangt 
so the only error is related to integration test configuration not yet being updated (agentskills signature has changed -> mock data is expecting different result) so none of the agentskills tests will work here.
Probably could defer that to a followup PR.

Since the `pyproject.toml` has a few updates, I think it'd be good to also update the poetry.lock file _just before_ this PR gets merged (Dependabot commits might happen in the meantime).

**Comment by ryanhoangt** [comment]
> Hello @ryanhoangt so the only error is related to integration test configuration not yet being updated (agentskills signature has changed -> mock data is expecting different result) so none of the agentskills tests will work here. Probably could defer that to a followup PR.
> 
> Since the `pyproject.toml` has a few updates, I think it'd be good to also update the poetry.lock file _just before_ this PR gets merged (Dependabot commits might happen in the meantime).

Got it, thanks a lot for your help 😄

**Comment by tobitege** [comment]
> > Hello @ryanhoangt so the only error is related to integration test configuration not yet being updated (agentskills signature has changed -> mock data is expecting different result) so none of the agentskills tests will work here. Probably could defer that to a followup PR.
> > Since the `pyproject.toml` has a few updates, I think it'd be good to also update the poetry.lock file _just before_ this PR gets merged (Dependabot commits might happen in the meantime).
> 
> Got it, thanks a lot for your help 😄 Actually let me also try to disable the `get_repomap` by default in agentskills to see if that fixes the intg tests

The issue with the integration tests is the changed docstring of AgentSkills, which I had requested earlier. I had assumed the test would be part of this PR in all its glory. 🤣 

I'd like to defer to the team, whether the tests can be deferred to a followup PR, so you could concentrate on the attributions.

**Comment by ramsey-coding** [comment]
I see you have copied over the `get_ranked_tags` from Aider to this repo. I dont get the logic of the ranking at all. I am wondering whether have you got the insight for the ranking? @neubig @ryanhoangt

**Comment by ryanhoangt** [comment]
> I see you have moved the `get_ranked_tags` function from Aider to this repository. I don't understand the logic of the ranking at all.
> 
> I'm wondering if any of you have insights into the ranking logic? @neubig @ryanhoangt @xingyaoww

Hi @ramsey-coding, we're planning to first integrate and test out the original implementation of `RepoMap` from `aider` with OpenDevin, so we won't make any modifications for now. If you want to dive deep into the ranking logic of `RepoMap`, I'd suggest reading the [blog](https://aider.chat/2023/10/22/repomap.html#optimizing-the-map) from `aider` and playing around with the source code.

**Comment by li-boxuan** [comment]
> none of the agentskills tests will work here. Probably could defer that to a followup PR

No we shouldn't defer it to a followup PR. The CI gate won't let you merge the PR anyways. You can ping me to help you fix the tests once you finish dev though.

**Comment by neubig** [comment]
Hey @ryanhoangt , please tell me when this is ready for another look! Happy to run the eval as necessary.

**Review by li-boxuan — COMMENTED** [review]
Can we please add this to `tests/unit/test_sandbox.py`?

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (neubig) [inline opendevin/runtime/plugins/agent_skills/agentskills.py:874]:
```suggestion
    messages: list[dict] = []  # FIXME: change to message history obtained from `EventStream` after
```

File: pyproject.toml
Comment (neubig) [inline pyproject.toml:71]: Are all of these dependencies necessary?

**Review by neubig — COMMENTED** [review]
Hey @ryanhoangt , I feel a bit bad saying this because I know you worked very hard on this, but given the level of complexity (and extra dependencies) that this code introduces, we might want to consider merging this in to main if it's increasing accuracy. Extra code and extra dependencies cause extra maintenance burden, so we do need to think about that a little.

As a next step, it may be good to debug the evaluation results. If you have them saved, we could potentially look at the trajectories and try to better understand what is going on with them.

File: pyproject.toml
Comment (ryanhoangt) [reply pyproject.toml:71]: Yes - it's required to pass the unit tests

**Pull request closed without merging**
